File: apps/packages/ui/src/utils/history-selection.ts
import { sha256 } from "@noble/hashes/sha2.js"
import { bytesToHex, utf8ToBytes } from "@noble/hashes/utils.js"
import type {
  CompareHistorySelectionV1, HistoryCursorV1, HistoryNodeV1, HistoryResolutionV1,
  HistorySelectedContentV1, HistorySelectionSnapshotV1, HistorySelectionV1, HistoryViewSelectionV1
} from "@/types/history-selection"

export class HistorySelectionError extends Error {
  constructor(readonly code: string) {
    super(code)
    this.name = "HistorySelectionError"
  }
}

const indexNodes = (nodes: readonly HistoryNodeV1[], validateGraph = false): Map<string, HistoryNodeV1> => {
  const byId = new Map<string, HistoryNodeV1>()
  let conversationId: string | undefined
  for (const node of nodes) {
    if (!node.id || byId.has(node.id)) throw new HistorySelectionError("duplicate_message_id")
    if (node.conversation_id) {
      if (conversationId && conversationId !== node.conversation_id) throw new HistorySelectionError("cross_conversation_parent")
      conversationId = node.conversation_id
    }
    byId.set(node.id, node)
  }
  if (validateGraph) {
    const resolved = new Set<string>()
    for (const node of nodes) {
      const walking = new Set<string>()
      let current: HistoryNodeV1 | undefined = node
      while (current && !resolved.has(current.id)) {
        if (walking.has(current.id)) throw new HistorySelectionError("cyclic_ancestry")
        walking.add(current.id)
        if (current.parent_id === null) break
        const parent: HistoryNodeV1 | undefined = byId.get(current.parent_id)
        if (!parent) throw new HistorySelectionError("missing_parent")
        if (current.conversation_id && parent.conversation_id !== current.conversation_id) {
          throw new HistorySelectionError("cross_conversation_parent")
        }
        current = parent
      }
      for (const id of walking) resolved.add(id)
    }
  }
  return byId
}

/** Resolve only explicit parent IDs. Source order, content and timestamps have no authority. */
export const resolveParentPath = (
  nodes: readonly HistoryNodeV1[], cursor: HistoryCursorV1
): readonly HistoryNodeV1[] => {
  const byId = indexNodes(nodes, true)
  if (cursor.kind === "empty") return []
  const selected = byId.get(cursor.message_id)
  if (!selected) throw new HistorySelectionError("missing_cursor")
  const path: HistoryNodeV1[] = []
  const visited = new Set<string>()
  let current: HistoryNodeV1 | undefined = selected
  while (current) {
    if (visited.has(current.id)) throw new HistorySelectionError("cyclic_ancestry")
    visited.add(current.id)
    path.push(current)
    if (current.parent_id === null) break
    const parent: HistoryNodeV1 | undefined = byId.get(current.parent_id)
    if (!parent) throw new HistorySelectionError("missing_parent")
    if (current.conversation_id && parent.conversation_id !== current.conversation_id) {
      throw new HistorySelectionError("cross_conversation_parent")
    }
    current = parent
  }
  path.reverse()
  return cursor.kind === "before_message" ? path.slice(0, -1) : path
}

/** A reviewed legacy order is immutable; original rows and parent edges remain intact. */
export const resolveLegacyProjection = (
  nodes: readonly HistoryNodeV1[], orderedPathIds: readonly string[], cursor: HistoryCursorV1
): readonly HistoryNodeV1[] => {
  const byId = indexNodes(nodes)
  const seen = new Set<string>()
  const path = orderedPathIds.map(id => {
    if (seen.has(id)) throw new HistorySelectionError("duplicate_projection_member")
    seen.add(id)
    const row = byId.get(id)
    if (!row) throw new HistorySelectionError("missing_projection_member")
    return row
  })
  if (cursor.kind === "empty") return []
  const at = orderedPathIds.indexOf(cursor.message_id)
  if (at < 0) throw new HistorySelectionError("missing_cursor")
  return path.slice(0, at + (cursor.kind === "after_message" ? 1 : 0))
}

/** Bind a separately loaded content payload to the exact captured path. */
export const bindSelectedHistoryContent = (
  rows: readonly HistoryNodeV1[], content: readonly HistorySelectedContentV1[]
): readonly HistorySelectedContentV1[] => {
  if (rows.length !== content.length || new Set(rows.map(row => row.id)).size !== rows.length ||
      rows.some((row, index) => row.id !== content[index]?.id || row.revision !== content[index]?.revision)) {
    throw new HistorySelectionError("selected_content_mismatch")
  }
  return Object.freeze(content.map(item => Object.freeze({
    id: item.id,
    revision: item.revision,
    message: item.message,
    images: Object.freeze([...item.images])
  })))
}

/** The nine-member tuple is the only normal-selection digest input. */
export const canonicalSelectionTuple = (selection: HistorySelectionV1): readonly unknown[] => [
  1,
  selection.owner_key,
  selection.conversation_id,
  [selection.interpretation.kind, selection.interpretation.kind === "legacy_linear_v1" ? selection.interpretation.projection_id : null],
  [selection.cursor.kind, selection.cursor.kind === "empty" ? null : selection.cursor.message_id],
  selection.purpose,
  selection.messages.map(({ id, revision }) => [id, revision]),
  selection.storage_context_digest,
  selection.request_context_digest
]

export const canonicalSelectionJson = (selection: HistorySelectionV1): string => JSON.stringify(canonicalSelectionTuple(selection))
export const selectionDigest = (selection: HistorySelectionV1): string => bytesToHex(sha256(utf8ToBytes(canonicalSelectionJson(selection))))

/** Comparison uses semantic source order, independent of stored cross-model parent edges. */
export const resolveComparisonProjection = (
  nodes: readonly HistoryNodeV1[], modelId: string, boundaryId: string
): readonly HistoryNodeV1[] => {
  indexNodes(nodes)
  const boundary = nodes.find(node => node.id === boundaryId)
  if (!boundary || !boundary.comparison || (!boundary.comparison.common && boundary.comparison.model_id !== modelId)) {
    throw new HistorySelectionError("invalid_comparison_boundary")
  }
  const end = nodes.indexOf(boundary)
  return nodes.slice(0, end + 1).filter(node => node.comparison?.common || node.comparison?.model_id === modelId)
}

export const canonicalComparisonTuple = (selection: CompareHistorySelectionV1): readonly unknown[] => [
  1, "comparison", selection.owner_key, selection.conversation_id, selection.model_id, selection.cluster_id,
  [selection.cursor.kind, selection.cursor.kind === "empty" ? null : selection.cursor.message_id],
  selection.messages.map(({ id, revision }) => [id, revision]),
  selection.storage_context_digest, selection.request_context_digest
]
export const canonicalComparisonJson = (selection: CompareHistorySelectionV1): string => JSON.stringify(canonicalComparisonTuple(selection))
export const comparisonDigest = (selection: CompareHistorySelectionV1): string => bytesToHex(sha256(utf8ToBytes(canonicalComparisonJson(selection))))

export const resolveHistorySelection = (
  snapshot: HistorySelectionSnapshotV1, view: HistoryViewSelectionV1, purpose: "send" | "fork", requestContextDigest: string
): HistoryResolutionV1 => {
  if (snapshot.version !== 1) return { status: "unsupported_history_capability", code: "unsupported_version" }
  if (snapshot.owner_key !== view.owner_key || snapshot.conversation_id !== view.conversation_id) {
    return { status: "invalid_history", code: "owner_conversation_mismatch" }
  }
  const status = snapshot.interpretation_status
  if (status.kind === "legacy_review_required") return { status: "legacy_review_required", code: "legacy_review_required" }
  if (status.kind !== view.interpretation.kind ||
      (status.kind === "legacy_linear_v1" && view.interpretation.kind === "legacy_linear_v1" && status.projection_id !== view.interpretation.projection_id)) {
    return { status: "stale_selection", code: "interpretation_mismatch" }
  }
  try {
    if (snapshot.nodes.some(node => node.conversation_id && node.conversation_id !== snapshot.conversation_id)) {
      throw new HistorySelectionError("cross_conversation_parent")
    }
    const rows = status.kind === "parent_graph_v1"
      ? resolveParentPath(snapshot.nodes, view.cursor)
      : resolveLegacyProjection(snapshot.nodes, status.ordered_path_ids, view.cursor)
    if (rows.some(row => !row.settled)) throw new HistorySelectionError("unsettled_message")
    const selection: HistorySelectionV1 = {
      version: 1, owner_key: snapshot.owner_key, conversation_id: snapshot.conversation_id,
      interpretation: view.interpretation, cursor: view.cursor, selection_revision: view.selection_revision,
      purpose, messages: rows.map(({ id, revision }) => ({ id, revision })), fences: snapshot.fences,
      storage_context_digest: snapshot.storage_context_digest, request_context_digest: requestContextDigest, selection_digest: ""
    }
    return { status: "ready", rows, selection: { ...selection, selection_digest: selectionDigest(selection) } }
  } catch (error) {
    if (!(error instanceof HistorySelectionError)) throw error
    return { status: error.code === "missing_cursor" ? "stale_selection" : "invalid_history", code: error.code }
  }
}

File: tldw_Server_API/app/core/Chat/history_selection.py
"""Pure, owner-neutral H1 history selection values and canonical encoding.

Owners validate storage context separately. A digest is provenance, not authority.
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from types import MappingProxyType
from typing import Any


@dataclass(frozen=True)
class HistorySelectionError(ValueError):
    """A stable code for malformed or unavailable ancestry."""

    code: str

    def __str__(self) -> str:
        return self.code


@dataclass(frozen=True)
class HistoryMessageRevisionV1:
    """Immutable accepted source member identity."""

    id: str
    revision: str


@dataclass(frozen=True)
class HistoryFencesV1:
    """Independent conversation, history and settings revisions."""

    conversation: str
    history: str
    settings: str


@dataclass(frozen=True)
class HistorySelectionV1:
    """Finalized selection submitted to an owner for admission."""

    version: int
    owner_key: str
    conversation_id: str
    interpretation: Mapping[str, str]
    cursor: Mapping[str, str]
    selection_revision: int
    purpose: str
    messages: tuple[HistoryMessageRevisionV1, ...]
    fences: HistoryFencesV1
    storage_context_digest: str
    request_context_digest: str
    selection_digest: str

    def __post_init__(self) -> None:
        """Detach caller-owned nested values before a selection can be admitted."""

        object.__setattr__(self, "interpretation", _freeze_json(self.interpretation))
        object.__setattr__(self, "cursor", _freeze_json(self.cursor))
        object.__setattr__(self, "messages", tuple(self.messages))


@dataclass(frozen=True)
class HistorySelectionSnapshotV1:
    """A complete owner-captured lightweight source manifest."""

    version: int
    owner_key: str
    conversation_id: str
    fences: HistoryFencesV1
    nodes: tuple[Mapping[str, Any], ...]
    source_digest: str
    interpretation_status: Mapping[str, Any]
    storage_context_digest: str

    def __post_init__(self) -> None:
        """Freeze the complete manifest and selected interpretation evidence."""

        object.__setattr__(self, "nodes", tuple(_freeze_json(row) for row in self.nodes))
        object.__setattr__(self, "interpretation_status", _freeze_json(self.interpretation_status))


@dataclass(frozen=True)
class HistoryAdmissionV1:
    """Owner-issued accepted input and immutable selected membership."""

    version: int
    owner_key: str
    conversation_id: str
    selection_digest: str
    messages: tuple[HistoryMessageRevisionV1, ...]
    input_message_id: str
    input_message_revision: str
    originating_selection_revision: int

    def __post_init__(self) -> None:
        object.__setattr__(self, "messages", tuple(self.messages))


def _freeze_json(value: Any) -> Any:
    """Copy H1's JSON-shaped captured fields into standard immutable containers."""

    if isinstance(value, Mapping):
        return MappingProxyType({key: _freeze_json(member) for key, member in value.items()})
    if isinstance(value, (list, tuple)):
        return tuple(_freeze_json(member) for member in value)
    return value


def _wire_json(value: Any) -> Any:
    """Convert captured JSON values to plain wire containers without deepcopy."""

    if isinstance(value, Mapping):
        return {key: _wire_json(member) for key, member in value.items()}
    if isinstance(value, tuple):
        return [_wire_json(member) for member in value]
    return value


def selection_to_wire(selection: HistorySelectionV1) -> dict[str, Any]:
    """Serialize a frozen core selection for strict Pydantic validation."""

    return {
        "version": selection.version,
        "owner_key": selection.owner_key,
        "conversation_id": selection.conversation_id,
        "interpretation": _wire_json(selection.interpretation),
        "cursor": _wire_json(selection.cursor),
        "selection_revision": selection.selection_revision,
        "purpose": selection.purpose,
        "messages": [{"id": member.id, "revision": member.revision} for member in selection.messages],
        "fences": {"conversation": selection.fences.conversation, "history": selection.fences.history, "settings": selection.fences.settings},
        "storage_context_digest": selection.storage_context_digest,
        "request_context_digest": selection.request_context_digest,
        "selection_digest": selection.selection_digest,
    }


def snapshot_to_wire(snapshot: HistorySelectionSnapshotV1) -> dict[str, Any]:
    """Serialize frozen manifest fields as ordinary JSON-shaped wire values."""

    return {
        "version": snapshot.version,
        "owner_key": snapshot.owner_key,
        "conversation_id": snapshot.conversation_id,
        "fences": {"conversation": snapshot.fences.conversation, "history": snapshot.fences.history, "settings": snapshot.fences.settings},
        "nodes": _wire_json(snapshot.nodes),
        "source_digest": snapshot.source_digest,
        "interpretation_status": _wire_json(snapshot.interpretation_status),
        "storage_context_digest": snapshot.storage_context_digest,
    }


def bind_selected_history_content(
    rows: Sequence[Mapping[str, Any]], content: Sequence[Mapping[str, Any]]
) -> tuple[Mapping[str, Any], ...]:
    """Reject content loaded at another source revision or in another order."""

    if len(rows) != len(content) or len({row["id"] for row in rows}) != len(rows) or any(
        row["id"] != item["id"] or row["revision"] != item["revision"]
        for row, item in zip(rows, content)
    ):
        raise HistorySelectionError("selected_content_mismatch")
    return tuple(
        _freeze_json({"id": item["id"], "revision": item["revision"], "message": item["message"], "images": item["images"]})
        for item in content
    )


def resolve_parent_path(
    nodes: Sequence[Mapping[str, Any]], cursor: Mapping[str, str]
) -> list[Mapping[str, Any]]:
    """Follow explicit parent IDs without sorting, content checks or mutation."""

    by_id: dict[str, Mapping[str, Any]] = {}
    conversation_id: str | None = None
    for row in nodes:
        row_id = row["id"]
        if not row_id or row_id in by_id:
            raise HistorySelectionError("duplicate_message_id")
        row_conversation = row.get("conversation_id")
        if row_conversation:
            if conversation_id and conversation_id != row_conversation:
                raise HistorySelectionError("cross_conversation_parent")
            conversation_id = row_conversation
        by_id[row_id] = row

    resolved: set[str] = set()
    for row in nodes:
        walking: set[str] = set()
        current: Mapping[str, Any] | None = row
        while current is not None and current["id"] not in resolved:
            if current["id"] in walking:
                raise HistorySelectionError("cyclic_ancestry")
            walking.add(current["id"])
            parent_id = current.get("parent_id")
            if parent_id is None:
                break
            parent = by_id.get(parent_id)
            if parent is None:
                raise HistorySelectionError("missing_parent")
            if current.get("conversation_id") and parent.get("conversation_id") != current["conversation_id"]:
                raise HistorySelectionError("cross_conversation_parent")
            current = parent
        resolved.update(walking)

    if cursor["kind"] not in ("empty", "before_message", "after_message"):
        raise HistorySelectionError("invalid_cursor")
    if cursor["kind"] == "empty":
        return []
    target = by_id.get(cursor["message_id"])
    if target is None:
        raise HistorySelectionError("missing_cursor")
    reversed_path: list[Mapping[str, Any]] = []
    visited: set[str] = set()
    current: Mapping[str, Any] = target
    while True:
        if current["id"] in visited:
            raise HistorySelectionError("cyclic_ancestry")
        visited.add(current["id"])
        reversed_path.append(current)
        parent_id = current.get("parent_id")
        if parent_id is None:
            break
        parent = by_id.get(parent_id)
        if parent is None:
            raise HistorySelectionError("missing_parent")
        if current.get("conversation_id") and parent.get("conversation_id") != current["conversation_id"]:
            raise HistorySelectionError("cross_conversation_parent")
        current = parent
    reversed_path.reverse()
    return reversed_path[:-1] if cursor["kind"] == "before_message" else reversed_path


def resolve_legacy_projection(
    nodes: Sequence[Mapping[str, Any]], ordered_path_ids: Sequence[str], cursor: Mapping[str, str]
) -> list[Mapping[str, Any]]:
    """Read reviewed membership without rewriting source rows or parent edges."""

    by_id = {row["id"]: row for row in nodes}
    if len(by_id) != len(nodes):
        raise HistorySelectionError("duplicate_message_id")
    conversations = {row.get("conversation_id") for row in nodes if row.get("conversation_id")}
    if len(conversations) > 1:
        raise HistorySelectionError("cross_conversation_parent")
    if len(set(ordered_path_ids)) != len(ordered_path_ids):
        raise HistorySelectionError("duplicate_projection_member")
    try:
        path = [by_id[row_id] for row_id in ordered_path_ids]
    except KeyError as exc:
        raise HistorySelectionError("missing_projection_member") from exc
    if cursor["kind"] not in ("empty", "before_message", "after_message"):
        raise HistorySelectionError("invalid_cursor")
    if cursor["kind"] == "empty":
        return []
    try:
        at = ordered_path_ids.index(cursor["message_id"])
    except ValueError as exc:
        raise HistorySelectionError("missing_cursor") from exc
    return path[: at + (cursor["kind"] == "after_message")]


def canonical_selection_tuple(selection: Mapping[str, Any]) -> list[Any]:
    """Return the exact nine-member H1 normal selection tuple."""

    interpretation = selection["interpretation"]
    cursor = selection["cursor"]
    return [
        1,
        selection["owner_key"],
        selection["conversation_id"],
        [interpretation["kind"], interpretation.get("projection_id")],
        [cursor["kind"], cursor.get("message_id")],
        selection["purpose"],
        [[member["id"], member["revision"]] for member in selection["messages"]],
        selection["storage_context_digest"],
        selection["request_context_digest"],
    ]


def canonical_selection_json(selection: Mapping[str, Any]) -> str:
    """Encode compact JSON with literal Unicode and no normalization."""

    return json.dumps(canonical_selection_tuple(selection), ensure_ascii=False, separators=(",", ":"))


def selection_digest(selection: Mapping[str, Any]) -> str:
    """Hash the canonical normal tuple as UTF-8 SHA-256."""

    return hashlib.sha256(canonical_selection_json(selection).encode("utf-8")).hexdigest()


def canonical_comparison_tuple(selection: Mapping[str, Any]) -> list[Any]:
    """Comparison tuple is tagged and independent of normal parent ancestry."""

    cursor = selection["cursor"]
    return [
        1, "comparison", selection["owner_key"], selection["conversation_id"],
        selection["model_id"], selection["cluster_id"],
        [cursor["kind"], cursor.get("message_id")],
        [[member["id"], member["revision"]] for member in selection["messages"]],
        selection["storage_context_digest"], selection["request_context_digest"],
    ]


def canonical_comparison_json(selection: Mapping[str, Any]) -> str:
    """Encode a comparison selection with the same Unicode rules."""

    return json.dumps(canonical_comparison_tuple(selection), ensure_ascii=False, separators=(",", ":"))


def comparison_digest(selection: Mapping[str, Any]) -> str:
    """Hash the tagged comparison tuple as UTF-8 SHA-256."""

    return hashlib.sha256(canonical_comparison_json(selection).encode("utf-8")).hexdigest()


def resolve_comparison_projection(
    nodes: Sequence[Mapping[str, Any]], model_id: str, boundary_id: str
) -> list[Mapping[str, Any]]:
    """Retain common rounds and one model in source order, ignoring cross-model edges."""

    seen: set[str] = set()
    conversation_id: str | None = None
    boundary_index: int | None = None
    for index, row in enumerate(nodes):
        row_id = row["id"]
        if not row_id or row_id in seen:
            raise HistorySelectionError("duplicate_message_id")
        seen.add(row_id)
        row_conversation = row.get("conversation_id")
        if row_conversation:
            if conversation_id and conversation_id != row_conversation:
                raise HistorySelectionError("cross_conversation_parent")
            conversation_id = row_conversation
        if row_id == boundary_id:
            comparison = row.get("comparison") or {}
            if not comparison or not (comparison.get("common") or comparison.get("model_id") == model_id):
                raise HistorySelectionError("invalid_comparison_boundary")
            boundary_index = index
    if boundary_index is None:
        raise HistorySelectionError("invalid_comparison_boundary")
    return [
        row for row in nodes[: boundary_index + 1]
        if row.get("comparison", {}).get("common") or row.get("comparison", {}).get("model_id") == model_id
    ]


def resolve_history_selection(
    snapshot: Mapping[str, Any], view: Mapping[str, Any], purpose: str, request_context_digest: str
) -> dict[str, Any]:
    """Resolve a coherent owner snapshot to a selection or structured non-ready state."""

    if snapshot["version"] != 1:
        return {"status": "unsupported_history_capability", "code": "unsupported_version"}
    if snapshot["owner_key"] != view["owner_key"] or snapshot["conversation_id"] != view["conversation_id"]:
        return {"status": "invalid_history", "code": "owner_conversation_mismatch"}
    status = snapshot["interpretation_status"]
    if status["kind"] == "legacy_review_required":
        return {"status": "legacy_review_required", "code": "legacy_review_required"}
    interpretation = view["interpretation"]
    if status["kind"] != interpretation["kind"] or (
        status["kind"] == "legacy_linear_v1" and status["projection_id"] != interpretation["projection_id"]
    ):
        return {"status": "stale_selection", "code": "interpretation_mismatch"}
    try:
        nodes = snapshot["nodes"]
        if any(row.get("conversation_id") and row["conversation_id"] != snapshot["conversation_id"] for row in nodes):
            raise HistorySelectionError("cross_conversation_parent")
        rows = (
            resolve_parent_path(nodes, view["cursor"])
            if status["kind"] == "parent_graph_v1"
            else resolve_legacy_projection(nodes, status["ordered_path_ids"], view["cursor"])
        )
        if any(not row["settled"] for row in rows):
            raise HistorySelectionError("unsettled_message")
        selection = {
            "version": 1,
            "owner_key": snapshot["owner_key"],
            "conversation_id": snapshot["conversation_id"],
            "interpretation": interpretation,
            "cursor": view["cursor"],
            "selection_revision": view["selection_revision"],
            "purpose": purpose,
            "messages": [{"id": row["id"], "revision": row["revision"]} for row in rows],
            "fences": snapshot["fences"],
            "storage_context_digest": snapshot["storage_context_digest"],
            "request_context_digest": request_context_digest,
        }
        selection["selection_digest"] = selection_digest(selection)
        return {"status": "ready", "selection": selection, "rows": rows}
    except HistorySelectionError as exc:
        return {"status": "stale_selection" if exc.code == "missing_cursor" else "invalid_history", "code": exc.code}
